spec/abci++/abci++_tmint_expected_behavior_002_draft.md

```diff
@@ -0,0 +1,184 @@
+  The number of time _p_'s Tendermint may call `RequestProcessProposal` with different proposed blocks for a given height is *unbounded*.
+  As a result, the Application may need to deal with an unbounded number of different proposals for a given height,
+  and also an unbounded number of candidate states if it is fully executing the proposed blocks upon `PrepareProposal` or `ProcessProposal`.
+  In order protects the processes' stability (memory, CPU), the Application has to:
+    * Be ready to discard candidate states if they become too many. In other words, the set of candidate states should be managed like a cache.
+    * Be able to execute the blocks upon `FinalizeBlock` if the block decided was one whose candidate state was discarded.
+
+### `ExtendVote`:  expectations from the Application
+
+>**TODO** (in different rounds). [Finish first discussion above]
+
+--
+
+If `ProcessProposal`'s outcome is _Reject_ for some proposed block. Tendermint guarantees that the block will not be the decision.
+
+--
+
+The validity of every transaction in a block (from the App's point of view), as well as the hashes in its header can be guaranteed if:
+
+* `ProcessProposal` *synchronously* handles every proposed block as though Tendermint had already decided on it.
+* All the properties of `ProcessProposal` and `FinalizeBlock` mentioned above hold.
+
+## [Points to discuss further]
+
+### Byzantine proposer
+
+>**TODO** [From Josef] We should understand the influence of equivocation on proposals in ABCI++
+>
+>N.B.#1: If Byzantine proposer proposes both A and B in the same round, today we might only receive A (proposal) and not B (due to p2p implementation)
+>
+>Sergio: some thoughts:
+>
+>[Thought 1] If Tendermint delivers all proposals from a Byzantine proposer _p_ in _h_ and _r_ to the App, we're vulnerable to DoS attacks,
+>as _p_ can send as many as it wishes:
+>
+>* Assuming N.B.#1 above gets "fixed" (A and B get delivered by p2p)
+>* Assuming the App fully executes the proposed block at _ProcessProposal_ time
+>
+>So, whenever N.B.#1 is "fixed" at p2p level, we need to ensure that only the the first proposal in round _r_, height _h_
+>gets delivered to the Application. Any subsequent proposal for that round should just be used to submit evidence, but not given to the App.
+>
+>[Thought 2] In terms of the consensus's safety properties, as far as pure equivocation on proposals is concerned,
+>I think we are OK as long as _f<n/3_.
+
+### `ProcessProposal`'s timeout (a.k.a. Zarko's Github comment)
+
+>**TODO** (to discuss): `PrepareProposal` is called synchronously. `ProcessProposal` may also want to fully process the block synchronously.
+>However, they stand on Tendermint's critical path, so the Tendermint's Propose timeout needs to accomodate that.
+>
+>Idea: Make propose timestamp (currently hardcoded to 3 secs in the Tendermint Go implementation) part of ConsensusParams,
```

Comment:
Mind elaborating on what this means?
Some possibly helpful information here:
https://github.com/tendermint/tendermint/issues/7274

Reply:
First of all, thanks for pointing me to the issue. I'll go through it when tackling this
This is actually one of the last points I need to address in the spec. Haven't had the time yet.
The point is that, for applications (e.g., same-block execution) that _must_ fully execute the block at `PrepareProposal` time, the time they spend executing the block is on the critical path, so "timeout propose" should be adapted to accound for this.
So what we are proposing here is to make this timeout a consensus param, so that the App (who knows best how long it takes to execute a block) can set it accordingly (and also change it if it sees it needs to be adapted).
Hopefully this is clear enough.

Reply:
Makes sense, and I'd be in favor of changing this to be a consensus parameters as well. I mostly wanted to note that this is actually not hardcoded, it's configurable in the validator's _local_ config.toml file.

Reply:
Following the ConsensusParams in yesterday's meeting, I've tackled this in a slightly different manner, so that ABCI++ can be released before soft upgrades